File: CODING/Code @ Java/5. Trees/Vertical_order_traversal_1.java
/* Program to get the vertical order of the Binary Tree.....

Time Complexity : O(n*k)....where 'k' is the range of Horizantal distance
Space Complexity : O(n)
*/
class Node
{
	int data;
	Node left, right, link;
	
	Node(int item)
	{
		data = item;
		
		left = right = link = null;
	}
}

class BinaryTree
{
	int min = 0, max = 0;
	
	void printElements( Node root, int order, int count)
	{
		if(root == null)
			return;
		
		if(count == order)
			System.out.print(root.data + " ");
		
		printElements(root.left, order, count-1);
		
		printElements(root.right, order, count+1);
	}
	
	void getMinMax( Node root, int vcount)
	{
		if(root == null)
			return;
		
		if(vcount < min)
			min = vcount;
		
		if(vcount > max)
			max = vcount;
		
		getMinMax(root.left, vcount-1);
		
		getMinMax(root.right, vcount+1);
	}
	
	void printVerticalOrder(Node root)
	{
		getMinMax(root, 0);
		
		for(int i=min; i<=max; ++i)
			printElements(root, i, 0);
	}
	
	public static void main(String args[])
	{
		BinaryTree ob = new BinaryTree();
		
		Node root = new Node(10);
		root.left = new Node(20);
		root.right = new Node(30);
		root.left.left = new Node(40);
		root.left.left.left = new Node(1);
		root.left.left.right = new Node(3);
		root.left.left.right.left = new Node(39);
		root.left.left.right.right = new Node(19);
		root.left.right = new Node(50);
		root.right.left = new Node(60);
		root.right.left.left = new Node(45);
		root.right.right = new Node(70);
		root.right.right.left = new Node(65);
		root.right.right.left.left = new Node(95);
		root.right.right.left.right = new Node(99);
		root.right.right.right = new Node(100);
		
		ob.printVerticalOrder(root);
	}
}

File: CODING/8. Trees/Vertical_order_traversal_1.c
/*
Program to get the vertical order of the Binary Tree.....

Time Complexity : O(n*k)....where 'k' is the range of Horizantal distance
Space Complexity : O(n)
*/
#include<stdio.h>
#include<stdlib.h>

struct node
{
	int data;
	struct node *left,*right;
};

struct node *newNode(int val)
{
	struct node *nn = (struct node*)malloc(sizeof(struct node));
	
	nn->data = val;
	
	nn->left = nn->right = NULL;

	return nn;
}

void PreOrder(struct node *root, int *MIN, int *MAX, int count)
{
	if(!root)
		return;
		
	*MIN = (count< *MIN) ? count : *MIN;

	*MAX = (count> *MAX) ? count : *MAX;

	PreOrder(root->left, MIN, MAX, count-1);

	PreOrder(root->right, MIN, MAX, count+1);
}

void printVerticalOrder(struct node *root, int MIN, int count)
{
	if(!root)
		return;
		
	if(count == MIN)
		printf(" %d,",root->data);

	printVerticalOrder(root->left, MIN, count-1);

	printVerticalOrder(root->right, MIN, count+1);
}

void main()
{
	int MIN = 0, MAX = 0;

	struct node *root = newNode(10);
	root->left = newNode(20);
	root->right = newNode(30);
	root->left->left = newNode(40);
	root->left->left->right = newNode(45);
	root->left->right = newNode(50);
	root->right->left = newNode(60);
	root->right->right = newNode(70);

	PreOrder(root, &MIN, &MAX, 0);

	for( ; MIN<=MAX; MIN++)
		printVerticalOrder(root,MIN,0);
}
